### PR TITLE
K8s: Add passive mode for query service

### PR DESCRIPTION
**Pull request by stephaniehingtgen**
**What is this feature?**

This PR adds a second client to the querier, `passiveModeClient`. If set, it will route the query as normal to the Grafana instance, and then kick off a go routine to send the request again to the passive mode client, which will send it to the datasource api server. It will then compare the responses to make sure they match.

**Which issue(s) does this PR fix?**:

Related to https://github.com/grafana/app-platform-wg/issues/123

### CODE REVIEW

**Review by charandas — APPROVED** [review]
LGTM!

**Comment by diegommm** [comment]
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

File: pkg/registry/apis/query/query.go [hunk 1]
```diff
@@ -217,17 +262,154 @@ func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req d
...
+	_, err := strconv.ParseInt(req.Request.TimeRange.From, 10, 64)
+	_, err2 := strconv.ParseInt(req.Request.TimeRange.To, 10, 64)
+	if (err != nil) && (err2 != nil) {
+		b.log.Debug("relative query - not running comparison")
+		return false
+	}
+
+	// Only compare Loki & Prometheus requests. We do not want to compare all requests
+	// as some datasources have costs attached to queries.
+	return req.PluginId == datasources.DS_PROMETHEUS || req.PluginId == datasources.DS_LOKI
```
Comment (andresmgot) [inline pkg/registry/apis/query/query.go:286]: another nit: It's probably faster to check this so I'd check the pluginId before parsing the time range

**Review by andresmgot — COMMENTED** [review]
I have only spotted a few nits, otherwise LGTM

File: pkg/registry/apis/query/query.go [hunk 2]
```diff
@@ -192,15 +195,57 @@ func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req d
...
+			rsp.Responses[q.RefID] = backend.DataResponse{
+				Status: backend.Status(code),
+				Error:  err,
+			}
+		}
+	}
+
+	if b.peformComparison(req) {
+		b.log.Debug("running passive mode comparison to multi-tenant")
+		go b.queryDataAndCompare(ctx, pluginID, pluginUID, comparisonReq, code, rsp, dur)
```
Comment (marefr) [inline pkg/registry/apis/query/query.go:239]: Nit. Would provide a contextual logger as argument here to get the `traceID` included in the log messages.

```go
ctxLogger := b.log.FromContext(ctx)
``` 

Maybe generally a good idea for any method in `queryAPIBuilder` part of a request.

File: pkg/registry/apis/query/query.go [hunk 3]
```diff
@@ -217,17 +262,154 @@ func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req d
...
+		// only compare duration if there is a response. Errors will usually be faster.
+		b.reportDurationDiff(pluginID, stDuration, elapsedTime)
+	}
+}
+
+func (b *QueryAPIBuilder) compareResults(pluginID string, stRsp, mtRsp *backend.QueryDataResponse) bool {
+	// If the length of both responses is different, we know that
+	// something is not right and can straight up return.
+	if len(stRsp.Responses) != len(mtRsp.Responses) {
+		b.log.Debug("response sizes are different", "expected", len(stRsp.Responses), "got", len(mtRsp.Responses))
```
Comment (marefr) [inline pkg/registry/apis/query/query.go:325]: Nit. For any log message of differences might be helpful to output more details about the request to be able to see what the query was, i.e. `data.QueryDataRequest`?

**Review by marefr — APPROVED** [review]
LGTM with some nits

File: pkg/registry/apis/query/query.go (hunk 1 above)
Comment (stephaniehingtgen) [reply pkg/registry/apis/query/query.go]: Good idea - updated!

File: pkg/registry/apis/query/query.go [hunk 4]
```diff
@@ -217,17 +262,154 @@ func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req d
...
 	}
 
-	// Create a response object with the error when missing (happens for client errors like 404)
-	if rsp == nil && err != nil {
-		rsp = &backend.QueryDataResponse{Responses: make(backend.Responses)}
-		for _, q := range req.Request.Queries {
-			rsp.Responses[q.RefID] = backend.DataResponse{
-				Status: backend.Status(code),
-				Error:  err,
+	return code, rsp, elapsedTime, err
```
[the remark this answers is not in the document]
Comment (stephaniehingtgen) [reply pkg/registry/apis/query/query.go]: Good idea - added :)

File: pkg/registry/apis/query/query.go [hunk 5]
```diff
@@ -192,15 +195,57 @@ func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req d
 	}
 
 	// Add user headers... here or in client.QueryData
+	pluginID := req.PluginId
+	pluginUID := req.UID
+	comparisonReq := datasourceRequest{
```
[the remark this answers is not in the document]
Comment (stephaniehingtgen) [reply pkg/registry/apis/query/query.go]: Good idea, modified!

File: pkg/registry/apis/query/query.go (hunk 2 above)
Comment (stephaniehingtgen) [reply pkg/registry/apis/query/query.go]: Ah good to know, thank you! Updated

**Review by andresmgot — APPROVED** [review]
LGTM

File: pkg/registry/apis/query/query.go [hunk 6]
```diff
@@ -192,15 +195,72 @@ func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req d
 	}
 
 	// Add user headers... here or in client.QueryData
+	pluginID := req.PluginId
+	pluginUID := req.UID
+	ctxLogger := b.log.FromContext(ctx)
+	if logger, ok := ctxLogger.(*log.ConcreteLogger); ok {
+		ctxLogger = log.WithPrefix(logger, "datasource", pluginID, "dsUID", pluginUID)
+	}
```
Comment (marefr) [inline pkg/registry/apis/query/query.go:203]:
```suggestion
	ctxLogger := b.log.New("datasource", pluginID, "dsUID", pluginUID).FromContext(ctx)
```

**Pull request closed without merging**